### ISSUE TITLE
[BUG] Characters "/*" detected as a comment with Grey color in shell scripts

### ISSUE THREAD

**Issue by pokemaniac05** [slapperwan/gh4a#1129]
I noticed that "/*" is considered as comment beginning in all files, including shell scripts, where comments start with "#". 
## How to reproduce this bug?
Open [this link](https://github.com/MatMoul/archfi/blob/master/archfi#L1080) in Octodroid and analyze highlighted line. From /* until EOF there's no */ and so this part is considered as comment by Octodroid syntax coloring engine and it becomes gray, without specific coloring for operators, commands,... 
## Useful information 
Here is a [screenshot](https://lh3.googleusercontent.com/fife/AAWUweVhjR6k0bjJIi9PKEmZopj9RzG1jjQqZ9bpzDy8Vc2WKISUtXJSK1EqLjMKWJe9MkTLgWN-OsOaWxb7RoCLYzHDxFW3szQOHoWTgag5XE2eZZpxAG5pXfYaI2veVc0A_8WOtcORthK-TNKrqXR-RZiQGBIlwQMJNAlQbDQoRZyomIlMpZqxZpEVocIjjAsxUXfsqf0vGn7wOkg0kTSgRZL1t1u0JpEmHOLy2iKHH9rNv6Z2o2tz6byjBINPWDJjpLlntRAV7Z5ZUR1GcyGImVZAPx8-qQWE5OHyigAd-cYELSyv-EJphU4ZwKTCtCUJoinYhZqYqUFRn8TIIfEnGmzl_oeDPV92MT6puSy7kE7LejYQilYLuj5YXCPa-gbhkaSRpFrFiOkLySzciQ8ILoX3i-C_2q73JIMI22Su0R3nCoQwIDcS07TMKeDnKDp_vMaZJ5OOEwoYAaUtLbwiJQAkcy2ubQUVt7rZFGIxjtzLxnnizkmXL11edsXPyxOEzHSa8CbiXpccJDJc3rl6EDOXskOieJ8r_6HuM90l8q8kKip0I1wdTefag00rvr1H3x9d4KmV77QFxtWxAohKNGBRN4_foz8T4QfQj1BxRlfSo9qs7xei5hqxHXBb9yD_MfJkLmsiZTj9bA6lKyM4KJ7-T82GYHbDvABCWcyMqkmXRmFNcecdCvggz1HYvhAtn5jVM1zk8eRAD8gGuEXEGXv8SMhsm0gHjP23bg=w1960-h3538-ft) of the shell script which triggered the bug (I viewed it on the hardware/software described below). 
I'm using Octodroid 4.6.7 on Android 10 EMUI 10.1.0 on Huawei P40 lite E and I used F-Droid to install and upgrade Octodroid. 
Device is not rooted and I'm using kernel 4.14.116.

## Suggestions
There might be an extension and/or shabang based syntax coloring engine, such as GNU nano's one or Vim's one.

**Comment by pokemaniac05** [comment]
Did someone read this issue ticket?

**Comment by pokemaniac05** [comment]
This bug is still present

**Comment by Fs00** [comment]
Hello @pokemaniac05, I've dug into the bug you reported and found out that the issue is more complex than it may seem.
The behavior you observed is not a bug in the highlighting of Shell scripts, but the normal behavior of the [syntax highlighting library](https://github.com/googlearchive/code-prettify) when no specific language is specified. I'll try to explain it in more detail.

The app chooses which syntax highlighting to apply by looking at the file extension. Since the file you linked has no extension, the app says to the highlighting library: _"I don't know the language of this code snippet, highlight it in the way you deem best"_ (if the file had a `.sh` extension, the app would have instead said to the library _"highlight this `sh` code snippet"_).
Therefore, in the case you reported, the highlighting library can't know the language of the code. In these cases, it applies a set of highlighting rules that mixes the characteristics of different languages (C, Java, Shell and some more) and that's why you see an almost correct Shell highlighting that however interprets `/*` as the start of a comment.

As you may have realized, there is no easy fix for this. We'd need to either (a) replace the syntax highlighting library with a "smarter" one or (b) add logic to the app to guess the language by peeking into the file contents, which looks a bit overkill to me.

**Comment by smichel17** [comment]
I was going to suggest that the best fix would be to open an issue with the library for smarter highlighting (peeking inside the file at the `#!`), but…

> This repository has been archived and is no longer maintained.

…so it's probably a good idea to move to a different library at some point *anyway*.

**Comment by pokemaniac05** [comment]
@Fs00 I would choose the second approach:
* App checks the extension and applies syntax highlighting theme
* If there's no extension app checks shabang, takes filename or (if the interpreter is `/usr/bin/env`) env parameter and checks it in an interpreter database (even wrote in json to be simple to analyze) that specifies language for every known program. 
* If the file has no extension and no shabang, it should default to POSIX shell script (like command line text editors)
For example, there's a python script called `bpytop` with shabang `#!/usr/bin/env python3`: app doesn't find extension and looks at first line, it's a shabang and cuts directory hierarchy, the interpreter is env and so the app takes first parameter (python3) and looks for it in shabang database, where python3 is referred to python language and finally the app applies python coloring scheme

**Comment by smichel17** [comment]
> every known program

Probably just shell, python, and maybe js would account for _most_ extension-less files.

> If the file has no extension and no shabang, it should default to POSIX shell script

I wonder if it would be a reasonable quick fix to do this instead of the current behavior, for all extension-less files.

**Comment by pokemaniac05** [comment]
There's also `cowsay` without extension that is written in perl

**Comment by Fs00** [comment]
I'm working on this right now and I've implemented a solution that chooses the highlighting language based on the content of the shebang line (something similar to the proposed "shebang database").
However, I've found that the Shell-specific syntax highlighting of the library is almost as broken as the "generic fallback" one. See the screenshot below, which is the linked "archfi" script above with Shell highlighting:

![archfi_OctoDroid_Debug](https://user-images.githubusercontent.com/30041551/147464961-93fed191-9122-4d06-acc5-5a15596e8187.png)

I've also tried to look at the extension-less Perl cowsay you mentioned above, and surprisingly the generic fallback highlighting behaves better than the Perl one.

**Perl-specific highlighting:**
![cowsay_OctoDroid_Debug](https://user-images.githubusercontent.com/30041551/147465026-407a27ff-7e79-4b54-b8ec-c0f9b02918f6.png)

**Generic highlighting:**
![cowsay_OctoDroid](https://user-images.githubusercontent.com/30041551/147465052-37725bde-c845-4171-8074-ade1540d804e.png)

For these reasons I'd drop the idea of shebang-based language detection for now, since it adds complexity without giving any real improvement.
The only way to fix this issue in a reasonable way would be to replace the highlighting library, which is something that I'm personally not interested in doing (fyi, I'm not the maintainer of the project).
